pkg/clustermesh/remote_cluster.go

```diff
@@ -115,52 +119,45 @@ func (rc *remoteCluster) getLogger() *logrus.Entry {
 	})
 }
 
-// releaseOldConnection releases the etcd connection to a remote cluster
-func (rc *remoteCluster) releaseOldConnection() {
-	rc.mutex.Lock()
-	ipCacheWatcher := rc.ipCacheWatcher
-	rc.ipCacheWatcher = nil
-
-	remoteNodes := rc.remoteNodes
-	rc.remoteNodes = nil
-
-	remoteIdentityCache := rc.remoteIdentityCache
-	rc.remoteIdentityCache = nil
-
-	remoteServices := rc.remoteServices
-	rc.remoteServices = nil
-
-	backend := rc.backend
-	rc.backend = nil
-	rc.mutex.Unlock()
+// releaseOldConnection releases the etcd connection to a remote cluster. Must
+// be called with rc.mutex held for writing.
+func (rc *remoteCluster) releaseOldConnection(parentCtx context.Context) {
+	if rc.ipCacheWatcher != nil {
+		rc.ipCacheWatcher.Close()
+		rc.ipCacheWatcher = nil
+	}
 
-	// Release resources asynchroneously in the background. Many of these
-	// operations may time out if the connection was closed due to an error
-	// condition.
-	go func() {
-		if ipCacheWatcher != nil {
-			ipCacheWatcher.Close()
-		}
-		if remoteNodes != nil {
-			remoteNodes.Close(context.TODO())
-		}
-		if remoteIdentityCache != nil {
-			remoteIdentityCache.Close()
-		}
-		if remoteServices != nil {
-			remoteServices.Close(context.TODO())
-		}
-		if backend != nil {
-			backend.Close()
-		}
-	}()
+	if rc.remoteNodes != nil {
+		ctx, cancel := context.WithTimeout(parentCtx, rc.releaseTimeout)
+		defer cancel()
+		rc.remoteNodes.Close(ctx)
+		rc.remoteNodes = nil
+	}
+	if rc.remoteIdentityCache != nil {
+		rc.remoteIdentityCache.Close()
+		rc.remoteIdentityCache = nil
+	}
+	if rc.remoteServices != nil {
+		ctx, cancel := context.WithTimeout(parentCtx, rc.releaseTimeout)
+		defer cancel()
+		rc.remoteServices.Close(ctx)
+		rc.remoteServices = nil
+	}
+	if rc.backend != nil {
+		rc.backend.Close()
+		rc.backend = nil
```

Comment:
If something is blocking, then it is in the `Close()` call here so any timeout needs to apply in there.

Quickly checking the various `Close()` calls that happen into etcd, they seem to be bound to the session context so we could `e.sessionCancel()` after the timeout to interrupt any etcd blocking operations.